File: algorithms/0491.increasing_subsequences/increasing_subsequences.cpp
class Solution {
public:
    void backtracking(const vector<int> &nums, int index) {
        if (sequence_.size() > 1) {
            subsequences_.push_back(sequence_);
        }
        unordered_set<int> uset;
        for (int i = index; i < nums.size(); ++i) {
            if ((!sequence_.empty() && sequence_.back() > nums[i])
                || uset.find(nums[i]) != uset.end()) {
                    continue;
            }
            uset.insert(nums[i]);
            sequence_.push_back(nums[i]);
            backtracking(nums, i + 1);
            sequence_.pop_back();
        }
    }
    vector<vector<int>> findSubsequences(vector<int>& nums) {
        backtracking(nums, 0);
        return subsequences_;
    }
private:
    vector<vector<int>> subsequences_;
    vector<int> sequence_;   
};

File: algorithms/0491.increasing_subsequences/increasing_subsequences.rs
use std::collections::HashSet;
impl Solution {
    pub fn find_subsequences(nums: Vec<i32>) -> Vec<Vec<i32>> {
        let mut subsequences = Vec::new();
        let mut sequence = Vec::new();
        Self::backtracking(&nums, 0, &mut sequence, &mut subsequences);
        subsequences
    }
    fn backtracking(nums: &Vec<i32>, index: usize, sequence: &mut Vec<i32>, subsequences: &mut Vec<Vec<i32>>) {
        if sequence.len() >= 2 {
            subsequences.push((*sequence).clone());
        }
        let mut bucket = HashSet::new();
        for i in index..nums.len() {
            if bucket.contains(&nums[i]) {
                continue;
            }
            if sequence.is_empty() || *sequence.last().unwrap() <= nums[i] {
                sequence.push(nums[i]);
                bucket.insert(nums[i]);
                Self::backtracking(nums, i + 1, sequence, subsequences);
                sequence.pop();
            }
        }
    }
}